sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java

```diff
@@ -0,0 +1,977 @@
+       * @param fhirStore the fhir store
+       * @param tempGcsPath the temp gcs path
+       * @param deadLetterGcsPath the dead letter gcs path
+       */
+      WriteBundlesToFilesFn(String fhirStore, String tempGcsPath, String deadLetterGcsPath) {
+        this.fhirStore = fhirStore;
+        this.tempGcsPath = tempGcsPath;
+        this.deadLetterGcsPath = deadLetterGcsPath;
+      }
+
+      /**
+       * Init client.
+       *
+       * @throws IOException the io exception
+       */
+      @Setup
+      public void initClient() throws IOException {
+        this.client = new HttpHealthcareApiClient();
+      }
+
+      /**
+       * Init batch.
+       *
+       * @throws IOException the io exception
+       */
+      @StartBundle
+      public void initFile() throws IOException {
+        // Write each bundle to newline delimited JSON file.
+        String filename = String.format("/fhirImportBatch-%s.ndjson", UUID.randomUUID().toString());
+        this.resourceId = FileSystems.matchNewResource(this.tempGcsPath + filename, false);
+        this.ndJsonChannel = FileSystems.create(resourceId, "application/ld+json");
+        if (mapper == null) {
+          this.mapper = new ObjectMapper();
+        }
+      }
+
+      /**
+       * Add to batch.
+       *
+       * @param context the context
+       * @throws IOException the io exception
+       */
+      @ProcessElement
+      public void addToFile(ProcessContext context, BoundedWindow window) throws IOException {
+        this.window = window;
+        HttpBody httpBody = context.element();
+        try {
+          // This will error if not valid JSON an convert Pretty JSON to raw JSON.
+          Object data = this.mapper.readValue(httpBody.getData(), Object.class);
+          String ndJson = this.mapper.writeValueAsString(data) + "\n";
+          this.ndJsonChannel.write(ByteBuffer.wrap(ndJson.getBytes(StandardCharsets.UTF_8)));
+        } catch (JsonProcessingException e) {
+          String resource =
+              String.format(
+                  "Failed to parse payload: %s as json at: %s : %s."
+                      + "Dropping message from batch import.",
+                  httpBody.toString(), e.getLocation().getCharOffset(), e.getMessage());
+          LOG.warn(resource);
+          context.output(
+              Write.FAILED_BODY, HealthcareIOError.of(httpBody, new IOException(resource)));
+        }
+      }
+
+      /**
+       * Close file.
+       *
+       * @param context the context
+       * @throws IOException the io exception
+       */
+      @FinishBundle
+      public void closeFile(FinishBundleContext context) throws IOException {
+        // Write the file with all elements in this bundle to GCS.
+        ndJsonChannel.close();
+        context.output(resourceId, window.maxTimestamp(), window);
+      }
+    }
+  }
+
+  /** The type Execute bundles. */
+  public static class ExecuteBundles extends PTransform<PCollection<HttpBody>, Write.Result> {
```

Comment:
Is there any specific reason that the input is HttpBody instead of String? It feels weird to me because we only use the data field in the API client.

Reply:
@lastomato
My original goal was just to mimic the client library as closely as possible to be familiar to users that are used to that interface and to smooth migration to use the client library (rather than using apache http clinet / requests as a workaround). 

For example, if the signature was PTransform<PCollection<String>, Write.Result> it might be confusing to a user if they need to base 64 encode the strings themselves or not.

I was also unsure if there would be some use cases where users would want to set certain other things like extensions or content type in the HttpBody. Though it's worth noting that in the current implementation we are hard coding content type and don't handle extensions.

If you feel strongly that the API should accept Strings I'm open to refactoring for that.

Reply:
discussed offline, will refactor to pass around as strings.